### PR TITLE
Pt 163817459 sc fsm check rounds

### PR DESCRIPTION
**Pull request by velzevur**
PT [SC: reject off-chain txs with erronous round](https://www.pivotaltracker.com/story/show/163817459)
This PR is based on top of PR #2129 so it waits for its merge

### CODE REVIEW

**Comment by velzevur** [comment]
Extedned the scope of this PR to cover the `state_hash` checks as well. Rebased on top of #2129 which based on latest `minerva`

**Comment by velzevur** [comment]
Rebased on latest `minerva`

**Comment by velzevur** [comment]
Rebased against `master`

File: apps/aechannel/src/aesc_offchain_state.erl [hunk 1]
```diff
@@ -97,23 +97,30 @@ is_latest_signed_tx(SignedTx, #state{signed_tx = LatestSignedTx}) ->
...
+            end,
+          fun() ->
+              case aesc_create_tx:state_hash(Tx) =:= hash(State) of
+                  true -> ok;
+                  false -> {error, bad_state_hash}
+              end
+          end],
+    lists:foldl(
+        fun(_, {error, _} = Err) -> Err;
+           (F, _) -> F()
```
Comment (lucafavatella) [inline apps/aechannel/src/aesc_offchain_state.erl]: Suggestion to make stricter - for readability - though not strictly needed.
```suggestion
           (F, ok) -> F()
```

File: apps/aechannel/src/aesc_offchain_state.erl [hunk 2]
```diff
@@ -132,41 +139,38 @@ check_reestablish_tx(SignedTx, State) ->
...
-    case Mod:round(TxI) - 1 of
-        0 when OldSignedTx == ?NO_TX ->
-            lager:debug("previous round = 0", []),
-            check_update_tx_(F, Mod, TxI, State, OnChainTrees, OnChainEnv, Opts);
-        PrevRound ->
+    case Mod of
+        aesc_close_mutual_tx -> ok; % no particular check for a close mutual
+        _ ->
+            PrevRound = Mod:round(TxI) - 1,
+            true = PrevRound > 0, % assert not channel_create
```
Comment (lucafavatella) [inline apps/aechannel/src/aesc_offchain_state.erl]: I checked that all code paths leading here specify tx type, and none is channel create. So comment on assertion makes sense. Ok.

File: apps/aechannel/src/aesc_offchain_state.erl [hunk 3]
```diff
@@ -132,41 +139,38 @@ check_reestablish_tx(SignedTx, State) ->
...
     lager:debug("Tx = ~p", [Tx]),
-    case Mod:round(TxI) - 1 of
-        0 when OldSignedTx == ?NO_TX ->
-            lager:debug("previous round = 0", []),
-            check_update_tx_(F, Mod, TxI, State, OnChainTrees, OnChainEnv, Opts);
-        PrevRound ->
+    case Mod of
+        aesc_close_mutual_tx -> ok; % no particular check for a close mutual
+        _ ->
+            PrevRound = Mod:round(TxI) - 1,
```
Comment (lucafavatella) [inline apps/aechannel/src/aesc_offchain_state.erl]: Based on code inspection, :round is exported by module.

File: apps/aechannel/test/aesc_fsm_SUITE.erl [hunk 4]
```diff
@@ -136,7 +144,46 @@ init_per_suite(Config) ->
...
+    ModRoundTooLow = fun(Round) -> Round end, % keep previous round
+    Config = init_per_group_(Config0),
+    [{wrong_create, fun(_, _) -> error(no_invalid_round_on_create) end},
+     {wrong_action, fun(C, P, M, F) -> wrong_round_action(C, P, M, F,
+                                                          ModRoundTooLow) end},
+     {wrong_action_detailed, fun(C, P, M, F, D) -> wrong_round_action(C, P, M, F,
+                                                          D, ModRoundTooLow) end}
+     | Config];
+init_per_group(round_too_high, Config0) ->
+    ModRoundTooLow = fun(Round) -> Round + 2 end, % skip one
```
Comment (lucafavatella) [inline apps/aechannel/test/aesc_fsm_SUITE.erl]: Misleading variable name `ModRoundTooLow`. Cosmetic.

**Comment by lucafavatella** [comment]
Labelling as WIP as needing rebase.

File: apps/aechannel/src/aesc_fsm.erl [hunk 5]
```diff
@@ -1964,11 +1967,17 @@ send_funding_created_msg(SignedTx, #data{channel_id = Ch,
 
 check_funding_created_msg(#{ temporary_channel_id := ChanId
                            , data                 := TxBin } = Msg,
-                          #data{ channel_id = ChanId} = Data) ->
+                          #data{ state = State, opts = Opts, 
+                                 channel_id = ChanId} = Data) ->
```
Comment (uwiger) [inline apps/aechannel/src/aesc_fsm.erl]: Missing space before `}`

File: apps/aechannel/test/aesc_fsm_SUITE.erl [hunk 6]
```diff
@@ -1043,27 +1053,124 @@ wrong_id_action(ChannelStuff, Poster, Malicious,
...
+                      {ok, Round} = aesc_fsm:get_round(Fsm),
+                      _Tx = update_tx(Tx0, set_round, [ModifyRoundFun(Round)])
+                  end,
+                  wrong_round).
+
+wrong_create_({I, R, #{initiator_amount := IAmt0, responder_amount := RAmt0,
+              push_amount := PushAmount} = Spec, Port, Debug},
+              Malicious, SignTxFun, ErrResponse) ->
+    Action = fun sign_signing_request/3,
+    TryCheaing =
```
Comment (uwiger) [inline apps/aechannel/test/aesc_fsm_SUITE.erl]: Should it be `TryCheating`?

**Comment by lucafavatella** [comment]
I see [system smoke tests failed on CI](https://circleci.com/gh/aeternity/aeternity/12073) with error unrelated:
```
%%% aest_sync_SUITE ==> new_node_joins_network: FAILED
%%% aest_sync_SUITE ==> {timeout,[{aest_nodes,wait_for_value,6,
                      [{file,"/home/circleci/aeternity/_build/system_test+test/extras/system_test/common/helpers/aest_nodes.erl"},
                       {line,812}]},
          {aest_nodes,loop_for_values,6,
                      [{file,"/home/circleci/aeternity/_build/system_test+test/extras/system_test/common/helpers/aest_nodes.erl"},
                       {line,774}]},
          {aest_sync_SUITE,new_node_joins_network,1,
                           [{file,"/home/circleci/aeternity/_build/system_test+test/extras/system_test/common/aest_sync_SUITE.erl"},
                            {line,237}]},
```

I see it is [tracked](https://www.pivotaltracker.com/story/show/163093465) as known intermittent test failure.

**Pull request merged**